tests/unit/models/test_interfaces.py

```diff
@@ -177,3 +178,36 @@ def _assert_data(self, dataset: Dataset) -> None:
     data = Dataset(tf.random.uniform([5, 7, 3]), tf.random.uniform([5, 7, 4]))
     stack.update(data)
     stack.optimize(data)
+
+
+def test_model_stack_reparam_sampler_raises_for_submodels_without_reparam_sampler() -> None:
+    model01 = _QuadraticModel([0.0, 0.5], [1.0, 0.3])
+    model2 = QuadraticMeanAndRBFKernel()
+    model_stack, _ = ModelStack((model01, 2), (model2, 1)), (model01, model2)  # type: ignore
+
+    with pytest.raises(NotImplementedError):
+        model_stack.reparam_sampler(1)
+
+
+def test_model_stack_reparam_sampler() -> None:
```

Comment:
(model stack would need more tests if you make the changes I suggested)

Reply:
Left changes for other PR (as described above)